Fix the MFX plugins loading issue on linux

nyanmisaka opened the pull request:
In the static build of MFX dispatcher on Linux, plugins are not shipped with dispatcher. Therefore, the ffmpeg linked to this dispatcher cannot find the location of plugins according to plugins.cfg, and causes HEVC encoding to fail.

So we should hardcode the default paths of some common plugins.cfg so that dispatcher can find plugins correctly.

Add a series of fallback paths for plugins.cfg, this should cover most linux distros.

Signed-off-by: nyanmisaka <nst799610810@gmail.com>

x264msna commented:
can it still respect MFX_PLUGINS_CONF_DIR for Linux ?
as it is now: it replaces MFX_PLUGINS_CONF_DIR  for hardcoded way.

maximd33 commented:
can it still respect MFX_PLUGINS_CONF_DIR for Linux ?
as it is now: it replaces MFX_PLUGINS_CONF_DIR for hardcoded way.

nyanmisaka commented:
> can it still respect MFX_PLUGINS_CONF_DIR for Linux ? as it is now: it replaces MFX_PLUGINS_CONF_DIR for hardcoded way.

MFX_PLUGINS_CONF_DIR is empty. So the dispatcher will always check `/plugins.cfg`.
Is it necessary to keep the current behavior?

maximd33 commented:
> MFX_PLUGINS_CONF_DIR is empty. So the dispatcher will always check `/plugins.cfg`. Is it necessary to keep the current behavior?

can be expected to be changed for the right handling, 
see :  https://github.com/Intel-Media-SDK/MediaSDK/blob/510d19dcace1d8c57567fdd40b557155ab11ab8e/api/mfx_dispatch/linux/CMakeLists.txt#L47

do you really need 3 hardcoded paths and in this order 
or CMakeLists.txt change will be more flexible (and will not break current behavior)?

lu-zero commented:
setting MFX_PLUGINS_CONF_DIR in autoconf and cmake maybe is a nicer place, either ways it sounds good to me.

BtbN commented on src/mfxloader.cpp:
```diff
@@ -452,7 +452,21 @@ mfxStatus MFXVideoUSER_Load(mfxSession session, const mfxPluginUID *uid, mfxU32
+                                         "/usr/local/share/mfx/plugins.cfg",
+                                         "/opt/intel/mediasdk/plugins/plugins.cfg" };
+        for (int i = 0; i < sizeof(plugins_paths) / sizeof(plugins_paths[0]); i++) {
+          if (fopen(plugins_paths[i], "r") != NULL) {
```
Doesn't this leave an open file descriptor dangling around?

nyanmisaka replied:
Fixed.

BtbN commented:
I think the existing path should just be the first in the new list that is checked. Removing it would be a breaking change, even though it always defaulting to / is a bit silly. (In before some odd embedded system started to rely on that...)

The reason for the change are portable static builds, which have to guess where the system installed its stuff to.

nyanmisaka committed:
Fix the MFX plugins loading issue on linux

add a series of fallback paths for plugins.cfg, this should cover most linux distros.

Signed-off-by: nyanmisaka <nst799610810@gmail.com>

nyanmisaka commented:
Setting MFX_PLUGINS_CONF_DIR is feasible, but the `parse()` function can only check the path once each time it is called.

Is there a better way to make `parse()` check for multiple paths in MFX_PLUGINS_CONF_DIR separated by colons?

maximd33 commented:
@nyanmisaka that will be ideal and most flexible 

if you need several paths....
